File: crates/stateset-embedded/src/price_schedules.rs
//! Price schedule operations (time-bounded pricing).

use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use stateset_core::{
    CreatePriceSchedule, PriceSchedule, PriceScheduleEntry, PriceScheduleFilter, PriceScheduleId,
    ProductId, Result, UpdatePriceSchedule,
};
use stateset_db::{Database, DatabaseCapability};
use std::sync::Arc;

/// Price schedule operations.
pub struct PriceSchedules {
    db: Arc<dyn Database>,
}

impl std::fmt::Debug for PriceSchedules {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("PriceSchedules").finish_non_exhaustive()
    }
}

impl PriceSchedules {
    pub(crate) fn new(db: Arc<dyn Database>) -> Self {
        Self { db }
    }

    /// Whether price schedules are supported by the active backend.
    #[must_use]
    pub fn is_supported(&self) -> bool {
        self.db.supports_capability(DatabaseCapability::PriceSchedules)
    }

    fn ensure(&self) -> Result<()> {
        self.db.ensure_capability(DatabaseCapability::PriceSchedules)
    }

    /// Create a new price schedule.
    pub fn create(&self, input: CreatePriceSchedule) -> Result<PriceSchedule> {
        self.ensure()?;
        self.db.price_schedules().create(input)
    }

    /// Get a price schedule by ID.
    pub fn get(&self, id: PriceScheduleId) -> Result<Option<PriceSchedule>> {
        self.ensure()?;
        self.db.price_schedules().get(id)
    }

    /// Update a price schedule.
    pub fn update(&self, id: PriceScheduleId, input: UpdatePriceSchedule) -> Result<PriceSchedule> {
        self.ensure()?;
        self.db.price_schedules().update(id, input)
    }

    /// List price schedules with optional filtering.
    pub fn list(&self, filter: PriceScheduleFilter) -> Result<Vec<PriceSchedule>> {
        self.ensure()?;
        self.db.price_schedules().list(filter)
    }

    /// Delete a price schedule and its entries.
    pub fn delete(&self, id: PriceScheduleId) -> Result<()> {
        self.ensure()?;
        self.db.price_schedules().delete(id)
    }

    /// Upsert a per-product scheduled price.
    pub fn set_entry(
        &self,
        id: PriceScheduleId,
        product_id: ProductId,
        price: Decimal,
    ) -> Result<PriceScheduleEntry> {
        self.ensure()?;
        self.db.price_schedules().set_entry(id, product_id, price)
    }

    /// Remove a per-product entry.
    pub fn delete_entry(&self, id: PriceScheduleId, product_id: ProductId) -> Result<()> {
        self.ensure()?;
        self.db.price_schedules().delete_entry(id, product_id)
    }

    /// List per-product entries for a schedule.
    pub fn list_entries(&self, id: PriceScheduleId) -> Result<Vec<PriceScheduleEntry>> {
        self.ensure()?;
        self.db.price_schedules().list_entries(id)
    }

    /// Resolve the effective scheduled price for a product at an instant.
    pub fn resolve_price(
        &self,
        product_id: ProductId,
        at: DateTime<Utc>,
    ) -> Result<Option<Decimal>> {
        self.ensure()?;
        self.db.price_schedules().resolve_price(product_id, at)
    }
}

File: cli/src/tools/price-schedules.js
/**
 * Price Schedule Tools Module
 *
 * MCP tool definitions for time-windowed scheduled pricing.
 */

import { z } from 'zod';
import { applyRequired } from '../utils/apply-guard.js';

const withPolicyDomain = (policyDomain, tools) => tools.map((tool) => ({ policyDomain, ...tool }));

export const priceScheduleTools = withPolicyDomain('price_schedules', [
  {
    name: 'check_price_schedules_supported',
    description: 'Check whether the price-schedules backend is available on this engine build.',
    inputSchema: {},
    permission: 'read',
    handler: async ({ commerce }) => {
      const supported = await commerce.priceSchedules.isSupported();
      return { success: true, supported };
    },
  },
  {
    name: 'list_price_schedules',
    description: 'List price schedules with optional filtering.',
    inputSchema: {
      isActive: z.boolean().optional().describe('Filter by active flag'),
      limit: z.number().int().min(1).optional().describe('Maximum results'),
      offset: z.number().int().min(0).optional().describe('Offset for pagination'),
    },
    permission: 'read',
    handler: async ({ commerce, params }) => {
      const priceSchedules = await commerce.priceSchedules.list({
        isActive: params.isActive,
        limit: params.limit,
        offset: params.offset,
      });
      return { success: true, count: priceSchedules.length, priceSchedules };
    },
  },
  {
    name: 'get_price_schedule',
    description: 'Get a price schedule by ID.',
    inputSchema: {
      priceScheduleId: z.string().min(1).describe('Price schedule ID'),
    },
    permission: 'read',
    handler: async ({ commerce, params }) => {
      const priceSchedule = await commerce.priceSchedules.get(params.priceScheduleId);
      if (!priceSchedule) {
        return { success: false, error: 'Price schedule not found' };
      }
      return { success: true, priceSchedule };
    },
  },
  {
    name: 'create_price_schedule',
    description: 'Create a price schedule.',
    inputSchema: {
      name: z.string().min(1).describe('Schedule name'),
      code: z.string().min(1).optional().describe('Optional schedule code'),
      currency: z.string().min(1).optional().describe('Currency code, e.g. "USD"'),
      startsAt: z.string().optional().describe('Start timestamp in ISO 8601'),
      endsAt: z.string().optional().describe('End timestamp in ISO 8601'),
      priority: z
        .number()
        .int()
        .optional()
        .describe('Priority used to break ties (higher wins); default 0'),
    },
    permission: 'write',
    handler: async ({ commerce, params, allowApply }) => {
      if (!allowApply) {
        return applyRequired('Create price schedule', params);
      }

      const priceSchedule = await commerce.priceSchedules.create({
        name: params.name,
        code: params.code,
        currency: params.currency,
        startsAt: params.startsAt,
        endsAt: params.endsAt,
        priority: params.priority,
      });
      return { success: true, message: 'Price schedule created', priceSchedule };
    },
  },
  {
    name: 'update_price_schedule',
    description: 'Update a price schedule.',
    inputSchema: {
      priceScheduleId: z.string().min(1).describe('Price schedule ID'),
      name: z.string().min(1).optional().describe('New name'),
      code: z.string().min(1).optional().describe('New code'),
      startsAt: z.string().optional().describe('New start timestamp in ISO 8601'),
      endsAt: z.string().optional().describe('New end timestamp in ISO 8601'),
      isActive: z.boolean().optional().describe('Active flag'),
      priority: z.number().int().optional().describe('New priority'),
    },
    permission: 'write',
    handler: async ({ commerce, params, allowApply }) => {
      if (!allowApply) {
        return applyRequired('Update price schedule', params);
      }

      const priceSchedule = await commerce.priceSchedules.update(params.priceScheduleId, {
        name: params.name,
        code: params.code,
        startsAt: params.startsAt,
        endsAt: params.endsAt,
        isActive: params.isActive,
        priority: params.priority,
      });
      return { success: true, message: 'Price schedule updated', priceSchedule };
    },
  },
  {
    name: 'delete_price_schedule',
    description: 'Delete a price schedule and its entries.',
    inputSchema: {
      priceScheduleId: z.string().min(1).describe('Price schedule ID'),
    },
    permission: 'write',
    handler: async ({ commerce, params, allowApply }) => {
      if (!allowApply) {
        return applyRequired('Delete price schedule', params);
      }

      await commerce.priceSchedules.delete(params.priceScheduleId);
      return { success: true, message: 'Price schedule deleted' };
    },
  },
  {
    name: 'set_price_schedule_entry',
    description: 'Upsert a per-product scheduled price on a price schedule.',
    inputSchema: {
      priceScheduleId: z.string().min(1).describe('Price schedule ID'),
      productId: z.string().min(1).describe('Product ID'),
      price: z.string().min(1).describe('Price as an exact decimal string, e.g. "19.99"'),
    },
    permission: 'write',
    handler: async ({ commerce, params, allowApply }) => {
      if (!allowApply) {
        return applyRequired('Set price schedule entry', params);
      }

      const entry = await commerce.priceSchedules.setEntry(
        params.priceScheduleId,
        params.productId,
        params.price,
      );
      return { success: true, message: 'Price schedule entry set', entry };
    },
  },
  {
    name: 'delete_price_schedule_entry',
    description: 'Remove a per-product entry from a price schedule.',
    inputSchema: {
      priceScheduleId: z.string().min(1).describe('Price schedule ID'),
      productId: z.string().min(1).describe('Product ID'),
    },
    permission: 'write',
    handler: async ({ commerce, params, allowApply }) => {
      if (!allowApply) {
        return applyRequired('Delete price schedule entry', params);
      }

      await commerce.priceSchedules.deleteEntry(params.priceScheduleId, params.productId);
      return { success: true, message: 'Price schedule entry deleted' };
    },
  },
  {
    name: 'list_price_schedule_entries',
    description: 'List per-product entries for a price schedule.',
    inputSchema: {
      priceScheduleId: z.string().min(1).describe('Price schedule ID'),
    },
    permission: 'read',
    handler: async ({ commerce, params }) => {
      const entries = await commerce.priceSchedules.listEntries(params.priceScheduleId);
      return { success: true, count: entries.length, entries };
    },
  },
  {
    name: 'resolve_scheduled_price',
    description:
      'Resolve the effective scheduled price for a product at an instant (defaults to now).',
    inputSchema: {
      productId: z.string().min(1).describe('Product ID'),
      at: z.string().optional().describe('Instant in ISO 8601 (defaults to now)'),
    },
    permission: 'read',
    handler: async ({ commerce, params }) => {
      const price = await commerce.priceSchedules.resolvePrice(params.productId, params.at);
      return { success: true, price, applies: price !== null };
    },
  },
]);

export default priceScheduleTools;
